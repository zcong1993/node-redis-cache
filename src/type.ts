// copy from https://github.com/sindresorhus/type-fest avoid TS2742

export type PromiseValue<
  PromiseType,
  Otherwise = PromiseType
> = PromiseType extends Promise<infer Value>
  ? { 0: PromiseValue<Value>; 1: Value }[PromiseType extends Promise<unknown>
      ? 0
      : 1]
  : Otherwise

type AsyncFunction = (...args: any[]) => Promise<unknown>

export type AsyncReturnType<Target extends AsyncFunction> = PromiseValue<
  ReturnType<Target>
>
