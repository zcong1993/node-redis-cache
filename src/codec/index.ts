export * from './codec'
export * from './json'
export * from './raw'
export * from './store'
