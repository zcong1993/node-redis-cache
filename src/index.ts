export * from './cache'
export * from './codec'
export * from './sharding'
