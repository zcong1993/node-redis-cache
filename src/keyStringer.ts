import { createHash } from 'crypto'

const maxKeyLen = 200

export type KeyStringer = (...args: any[]) => string

export const md5KeyStringer: KeyStringer = (...args: any[]) => {
  if (args.length === 0) {
    return ''
  }
  return createHash('md5').update(JSON.stringify(args)).digest('hex')
}

const isSimpleType = (v: any) => {
  return ['string', 'number', 'boolean'].includes(typeof v)
}

/**
 * combineKeyStringer will use origin args (array.join) to build the key when
 *   1. all args is simple type [string, number, boolean]
 *   2. output key length less than 200
 * otherwise it will fallback to md5KeyStringer
 */
export const combineKeyStringer: KeyStringer = (...args: any[]) => {
  if (args.some((v) => !isSimpleType(v))) {
    return md5KeyStringer(...args)
  }

  const key = args.join('|')
  if (key.length > maxKeyLen) {
    return md5KeyStringer(...args)
  }

  return key
}
