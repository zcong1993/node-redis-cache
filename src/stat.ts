export interface Stat {
  requestsCounter: number
  hitCounter: number
  errorsCounter: number
  hitNotFoundCacheCounter: number
}

export const createStat = (): Stat => {
  return {
    requestsCounter: 0,
    hitCounter: 0,
    errorsCounter: 0,
    hitNotFoundCacheCounter: 0,
  }
}
