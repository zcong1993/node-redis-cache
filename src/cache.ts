import { Redis } from 'ioredis'
import { createHash } from 'crypto'
import { Singleflight } from '@zcong/singleflight'
import { getCodec } from './codec'

export interface Option {
  redis: Redis
  prefix: string
  codec?: string
}

export type Hasher = (...args: any[]) => string

export const md5Hasher: Hasher = (...args: any[]) => {
  return createHash('md5').update(JSON.stringify(args)).digest('hex')
}

export interface ErrorEvent {
  key: string
  error: Error
  action: string
}

export type ErrorHandler = (errorEvent: ErrorEvent) => void
const noopHandler: ErrorHandler = () => {}

export class RedisCache {
  onError: ErrorHandler = noopHandler
  private readonly sf = new Singleflight()
  constructor(private readonly option: Option) {
    if (!this.option.codec) {
      this.option.codec = 'json'
    }
  }

  async cacheFn<T = any>(
    key: string,
    fn: () => Promise<T>,
    expire: number,
    codec?: string
  ): Promise<T> {
    try {
      const cached = await this.get(key, codec)
      if (cached !== null) {
        return cached
      }
    } catch (err) {
      this.onError({
        key,
        error: err,
        action: 'get cache',
      })
    }

    return this.sf.do(key, async () => {
      const res = await fn()
      try {
        await this.set(key, res, expire, codec)
      } catch (err) {
        this.onError({
          key,
          error: err,
          action: 'set cache',
        })
      }
      return res
    })
  }

  cacheWrapper<T extends (...args: any[]) => Promise<any>>(
    keyPrefix: string,
    fn: T,
    expire: number,
    codec?: string,
    keyHasher: Hasher = md5Hasher
  ): T {
    return (((...args: any[]) => {
      const cacheKey = RedisCache.joinKey(keyPrefix, keyHasher(...args))
      return this.cacheFn(cacheKey, () => fn(...args), expire, codec)
    }) as any) as T
  }

  async get<T = any>(key: string, codec?: string): Promise<T> {
    const val = await this.option.redis.get(this.buildKey(key))
    if (!val) {
      return null
    }
    return this.getCodecByName(codec).decode(val)
  }

  async set<T = any>(key: string, val: T, expire: number, codec?: string) {
    const vv = this.getCodecByName(codec).encode(val)
    await this.option.redis.set(this.buildKey(key), vv, 'ex', expire)
  }

  async delete(key: string) {
    await this.option.redis.del(this.buildKey(key))
  }

  private buildKey(key: string) {
    return RedisCache.joinKey(this.option.prefix, key)
  }

  private getCodecByName(codecName?: string) {
    return getCodec(codecName ?? this.option.codec)
  }

  private static joinKey(...keys: string[]) {
    return keys.join(':')
  }
}
