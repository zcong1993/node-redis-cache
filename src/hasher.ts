import { KeyStringer, md5KeyStringer } from './keyStringer'

/**
 * @deprecated use KeyStringer
 */
export type Hasher = KeyStringer
/**
 * @deprecated use md5KeyStringer
 */
export const md5Hasher = md5KeyStringer
