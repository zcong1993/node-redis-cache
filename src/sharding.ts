import * as HashRing from 'hashring'
import type { Servers } from 'hashring'
import { AsyncReturnType } from 'type-fest'
import { RedisCache, Cacher } from './cache'
import { bindThis, debug } from './utils'
import { Hasher, md5Hasher } from './hasher'

export interface Node {
  key: string
  redisCache: RedisCache
  weight: number
}

export interface ShardingOption {
  nodes: Node[]
}

export class ShardingCache implements Cacher {
  private nodeMap = new Map<string, RedisCache>()
  private ring: HashRing

  constructor(option: ShardingOption) {
    const servers: Servers = {}
    for (const node of option.nodes) {
      this.nodeMap.set(node.key, node.redisCache)
      servers[node.key] = { weight: node.weight }
    }
    this.ring = new HashRing(servers)
  }

  async cacheFn<F extends () => Promise<unknown>>(
    key: string,
    fn: F,
    expire: number,
    codec?: string,
    thisArg?: ThisParameterType<F>
  ): Promise<AsyncReturnType<F>> {
    const node = this.pickNode(key)
    return node.cacheFn(key, fn, expire, codec, thisArg)
  }

  cacheWrapper<T extends (...args: any[]) => Promise<any>>(
    keyPrefix: string,
    fn: T,
    expire: number,
    codec?: string,
    keyHasher: Hasher = md5Hasher,
    thisArg?: ThisParameterType<T>
  ): T {
    return (((...args: any[]) => {
      const cacheKey = RedisCache.joinKey(keyPrefix, keyHasher(...args))
      return this.cacheFn(
        cacheKey,
        () => bindThis(fn, thisArg)(...args),
        expire,
        codec
      )
    }) as any) as T
  }

  async deleteFnCache(
    keyPrefix: string,
    args: any[],
    keyHasher: Hasher = md5Hasher
  ) {
    const cacheKey = RedisCache.joinKey(keyPrefix, keyHasher(...args))
    await this.delete(cacheKey)
  }

  async get<T = any>(key: string, codec?: string): Promise<[T, boolean]> {
    const node = this.pickNode(key)
    return node.get(key, codec)
  }

  async set<T = any>(key: string, val: T, expire: number, codec?: string) {
    const node = this.pickNode(key)
    await node.set(key, val, expire, codec)
  }

  async delete(...keys: string[]) {
    await Promise.all(
      keys.map((key) => {
        const node = this.pickNode(key)
        return node.delete(key)
      })
    )
  }

  async clean(match: string = '*', count: number = 100) {
    await Promise.all(
      [...this.nodeMap.values()].map((node) => node.clean(match, count))
    )
  }

  private pickNode(key: string) {
    const nodeKey = this.ring.get(key)
    debug(`pick node, key: ${key}, nodeKey: ${nodeKey}`)
    return this.nodeMap.get(nodeKey)
  }
}
