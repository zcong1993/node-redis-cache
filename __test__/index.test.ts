import * as Redis from 'ioredis'
import { ValueType } from 'ioredis'
import { register } from 'prom-client'
import {
  RedisCache,
  Codec,
  registerCodec,
  setDefaultCodec,
  getCodec,
} from '../src'
import { createStat } from '../src/stat'

const mockFn = async <T>(n: number, val: T) =>
  new Promise((r) => {
    setTimeout(() => r(val), n)
  })

const repeatCall = (n: number, fn: Function) =>
  Promise.all(
    Array(n)
      .fill(null)
      .map((_) => fn())
  )

const redis = new Redis()

beforeAll(async () => {
  await redis.flushdb()
})

afterAll(() => {
  redis.disconnect()
}, 1000)

it('cacheWrapper should works well', async () => {
  const fn = jest.fn(mockFn)
  const c = new RedisCache({
    redis,
    prefix: 'test',
  })

  const mockRes = { name: 'test', age: 18 }
  const cf = c.cacheWrapper('fn', fn, 5)
  expect(await cf(100, mockRes)).toEqual(mockRes)
  expect(fn).toBeCalledTimes(1)
  await repeatCall(10, async () => {
    const res = await cf(100, mockRes)
    expect(res).toEqual(mockRes)
  })
  expect(fn).toBeCalledTimes(1)

  const mockRes2 = { name: 'test2', age: 19 }
  expect(await cf(100, mockRes2)).toEqual(mockRes2)
  expect(fn).toBeCalledTimes(2)
  await repeatCall(10, async () => {
    const res = await cf(100, mockRes2)
    expect(res).toEqual(mockRes2)
  })
  expect(fn).toBeCalledTimes(2)
})

it('not found should cache', async () => {
  const fn = jest.fn(mockFn)
  const c = new RedisCache({
    redis,
    prefix: 'test1',
  })

  const cf = c.cacheWrapper('fn', fn, 5)
  expect(await cf(100, null)).toBe(null)
  expect(fn).toBeCalledTimes(1)
  await repeatCall(10, () => cf(100, null))
  expect(fn).toBeCalledTimes(1)
})

it('raw codec should works well', async () => {
  const fn = jest.fn(mockFn)
  const c = new RedisCache({
    redis,
    prefix: 'test2',
    codec: 'raw',
  })

  const mockRes = 'raw test'
  const cf = c.cacheWrapper('fn', fn, 5)
  expect(await cf(100, mockRes)).toBe(mockRes)
  expect(fn).toBeCalledTimes(1)
  await repeatCall(10, async () => {
    const res = await cf(100, mockRes)
    expect(res).toBe(mockRes)
  })
  expect(fn).toBeCalledTimes(1)
})

it('singleflight should works well', async () => {
  const fn = jest.fn(mockFn)
  const c = new RedisCache({
    redis,
    prefix: 'test3',
  })

  const mockRes = { name: 'test', age: 18 }
  const cf = c.cacheWrapper('fn', fn, 5)
  await repeatCall(10, async () => {
    const res = await cf(100, mockRes)
    expect(res).toEqual(mockRes)
  })
  expect(fn).toBeCalledTimes(1)
})

it('cacheFn should works well', async () => {
  const fn = jest.fn(mockFn)
  const c = new RedisCache({
    redis,
    prefix: 'test4',
  })

  const mockRes = { name: 'test', age: 18 }
  const cf = () => c.cacheFn('fn', () => fn(100, mockRes), 5)
  expect(await cf()).toEqual(mockRes)
  await repeatCall(10, async () => {
    const res = await cf()
    expect(res).toEqual(mockRes)
  })
  expect(fn).toBeCalledTimes(1)
})

it('bindThis should works well', async () => {
  const fn = jest.fn(mockFn)
  const c = new RedisCache({
    redis,
    prefix: 'test41',
  })

  const mockRes = { name: 'test', age: 18 }

  class T {
    async test() {
      return this.testThis(100, mockRes)
    }

    async test2(...args: Parameters<typeof mockFn>) {
      return this.testThis(...args)
    }

    async testThis(...args: Parameters<typeof mockFn>) {
      return fn(...args)
    }
  }

  const t = new T()

  const cf = () => c.cacheFn('fn', t.test, 5, undefined, t)
  expect(await cf()).toEqual(mockRes)
  await repeatCall(10, async () => {
    const res = await cf()
    expect(res).toEqual(mockRes)
  })
  expect(fn).toBeCalledTimes(1)

  const cf1 = c.cacheWrapper('fn2', t.test2, 5, undefined, undefined, t)
  await repeatCall(10, async () => {
    const res = await cf1(100, mockRes)
    expect(res).toEqual(mockRes)
  })
  expect(fn).toBeCalledTimes(2)
})

it('custom codec should works well', async () => {
  const testCodecName = 'testCodec'
  class TestCodec implements Codec {
    name() {
      return testCodecName
    }

    encode<T = any>(data: T): ValueType {
      if (typeof data !== 'object') {
        throw new Error('only support object')
      }
      const tmp: any[] = []
      for (const [k, v] of Object.entries(data)) {
        tmp.push(`${k}-${v}`)
      }
      return tmp.join('|')
    }

    decode<T>(val: ValueType): T {
      const vv = val as string
      const res: any = {}
      const tmpArr = vv.split('|')
      tmpArr.forEach((el) => {
        const tmp = el.split('-')
        res[tmp[0]] = tmp[1]
      })
      return res
    }
  }

  registerCodec(new TestCodec())
  setDefaultCodec(testCodecName)

  const fn = jest.fn(mockFn)
  const c = new RedisCache({
    redis,
    prefix: 'test5',
  })

  const mockRes = { name: 'test' }
  const cf = () => c.cacheFn('fn', () => fn(100, mockRes), 5)
  expect(await cf()).toEqual(mockRes)
  await repeatCall(10, async () => {
    const res = await cf()
    expect(res).toEqual(mockRes)
  })
  expect(fn).toBeCalledTimes(1)

  expect(() => setDefaultCodec(testCodecName)).toThrow()
  expect(() => getCodec('notExists')).toThrow()
  expect(getCodec(testCodecName).name()).toBe(testCodecName)
  expect(getCodec().name()).toBe(testCodecName)
})

it('withPrometheus should works well', async () => {
  const fn = jest.fn(mockFn)
  const c = new RedisCache({
    redis,
    prefix: 'test6',
    withPrometheus: true,
    name: 'my-cache1',
  })

  const cf = c.cacheWrapper('fn', fn, 5)
  expect(await cf(100, null)).toBe(null)
  expect(fn).toBeCalledTimes(1)
  await repeatCall(10, () => cf(100, null))
  expect(fn).toBeCalledTimes(1)

  expect(await register.metrics()).toMatchSnapshot()
  expect(c.getStat()).toMatchSnapshot('stat')

  const mockRes2 = { name: 'test2', age: 19 }
  expect(await cf(100, mockRes2)).toEqual(mockRes2)
  expect(fn).toBeCalledTimes(2)
  await repeatCall(10, async () => {
    const res = await cf(100, mockRes2)
    expect(res).toEqual(mockRes2)
  })
  expect(fn).toBeCalledTimes(2)

  expect(await register.metrics()).toMatchSnapshot()
  expect(c.getStat()).toMatchSnapshot('stat 2')

  c.resetStat()
  expect(c.getStat()).toEqual(createStat())
})

it('deleteFnCache should works well', async () => {
  const fn = jest.fn(mockFn)
  const c = new RedisCache({
    redis,
    prefix: 'test7',
  })

  const mockRes = { name: 'test', age: 18 }
  const cf = c.cacheWrapper('fn', fn, 5)
  expect(await cf(100, mockRes)).toEqual(mockRes)
  expect(fn).toBeCalledTimes(1)
  await repeatCall(2, async () => {
    const res = await cf(100, mockRes)
    expect(res).toEqual(mockRes)
  })
  expect(fn).toBeCalledTimes(1)

  await c.deleteFnCache('fn', [100, mockRes])

  expect(await cf(100, mockRes)).toEqual(mockRes)
  expect(fn).toBeCalledTimes(2)
})

export async function testClean(
  c: RedisCache,
  redis: any,
  isCluster: boolean = false
) {
  const fn = jest.fn(mockFn)
  const cf = c.cacheWrapper('fn', fn, 5)

  const getDbSize = async () => {
    if (!isCluster) {
      return redis.dbsize()
    }

    let total = 0
    for (const node of redis.nodes('master')) {
      total += await node.dbsize()
    }
    return total
  }

  for (let i = 0; i < 10; i++) {
    const mockRes = { name: 'test', age: i }
    await cf(100, mockRes)
  }

  expect(await getDbSize()).toBe(10)
  await c.clean()
  expect(await getDbSize()).toBe(0)

  await redis.set('aa', 'aa', 'ex', 10)

  for (let i = 0; i < 10; i++) {
    const mockRes = { name: 'test', age: i }
    await cf(100, mockRes)
  }

  expect(await getDbSize()).toBe(11)
  await c.clean()
  expect(await getDbSize()).toBe(1)
  await redis.del('aa')

  await c.set('p1:aaa', 'test', 5, 'raw')
  await c.set('p1:bbb', 'test', 5, 'raw')
  await c.set('p2:aaa', 'test', 5, 'raw')
  expect(await getDbSize()).toBe(3)
  await c.clean('p1:*', 10)
  expect(await getDbSize()).toBe(1)
}

it('clean should works well', async () => {
  const redis = new Redis('redis://localhost/5')
  const c = new RedisCache({
    redis,
    prefix: 'test8',
  })

  await testClean(c, redis)
})

it('should auto delete invalid cache', async () => {
  const redis = new Redis('redis://localhost/7')
  const c = new RedisCache({
    redis,
    prefix: 'test10',
  })

  const fn = jest.fn()
  c.onError = fn

  const realKey = RedisCache.joinKey('test10', 'aaa')
  await redis.set(realKey, 'xxx', 'ex', 10)
  const res = await c.get('aaa')
  expect(res).toEqual([null, false])

  expect(await redis.get(realKey)).toBeNull()
  expect(fn).toBeCalledTimes(1)
})
