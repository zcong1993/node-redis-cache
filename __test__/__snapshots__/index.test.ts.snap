// Vitest Snapshot v1

exports[`withPrometheus should works well > stat 1`] = `
{
  "errorsCounter": 0,
  "hitCounter": 10,
  "hitNotFoundCacheCounter": 10,
  "requestsCounter": 11,
}
`;

exports[`withPrometheus should works well > stat 2 1`] = `
{
  "errorsCounter": 0,
  "hitCounter": 20,
  "hitNotFoundCacheCounter": 10,
  "requestsCounter": 22,
}
`;

exports[`withPrometheus should works well 1`] = `
"# HELP node_cache_requests_total Total number of requests to the cache.
# TYPE node_cache_requests_total counter
node_cache_requests_total{name=\\"my-cache1\\"} 11

# HELP node_cache_hits_total Total number of requests to the cache that were a hit.
# TYPE node_cache_hits_total counter
node_cache_hits_total{name=\\"my-cache1\\"} 10

# HELP node_cache_errors_total Total number of errors to the cache.
# TYPE node_cache_errors_total counter

# HELP node_cache_hit_not_found_cache_total Total number of requests to the cache that hit a not found cache.
# TYPE node_cache_hit_not_found_cache_total counter
node_cache_hit_not_found_cache_total{name=\\"my-cache1\\"} 10
"
`;

exports[`withPrometheus should works well 2`] = `
"# HELP node_cache_requests_total Total number of requests to the cache.
# TYPE node_cache_requests_total counter
node_cache_requests_total{name=\\"my-cache1\\"} 22

# HELP node_cache_hits_total Total number of requests to the cache that were a hit.
# TYPE node_cache_hits_total counter
node_cache_hits_total{name=\\"my-cache1\\"} 20

# HELP node_cache_errors_total Total number of errors to the cache.
# TYPE node_cache_errors_total counter

# HELP node_cache_hit_not_found_cache_total Total number of requests to the cache that hit a not found cache.
# TYPE node_cache_hit_not_found_cache_total counter
node_cache_hit_not_found_cache_total{name=\\"my-cache1\\"} 10
"
`;

exports[`withPrometheus should works well: stat 1`] = `
Object {
  "errorsCounter": 0,
  "hitCounter": 10,
  "hitNotFoundCacheCounter": 10,
  "requestsCounter": 11,
}
`;

exports[`withPrometheus should works well: stat 2 1`] = `
Object {
  "errorsCounter": 0,
  "hitCounter": 20,
  "hitNotFoundCacheCounter": 10,
  "requestsCounter": 22,
}
`;
